cpp/src/cylon/net/ucx/ucx_channel.cpp

```diff
@@ -0,0 +1,500 @@
+ * @return unsigned long int -  the compound tag
+ */
+static unsigned long int getTag(int edge,
+                  int sender) {
+  return (((unsigned long int)edge)<<32) + sender;
+}
+
+/**
+ * UCX Receive
+ * Modeled after the IRECV function of MPI
+ * @param [out] buffer - Pointer to the output buffer
+ * @param [in] count - Size of the receiving data
+ * @param [in] sender - MPI id of the sender
+ * @param [out] ctx - ucx::ucxContext object, used for tracking the progress of the request
+ */
+void UCXChannel::UCX_Irecv(void *buffer,
+                            size_t count,
+                            int sender,
+                            ucx::ucxContext* ctx) {
+  // To hold the status of operations
+  ucs_status_ptr_t status;
+
+  ucp_request_param_t recvParam;
+  recvParam.op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK |
+      UCP_OP_ATTR_FIELD_USER_DATA |
+      UCP_OP_ATTR_FLAG_NO_IMM_CMPL;
+  recvParam.cb.recv = recvHandler;
+  recvParam.user_data = ctx;
+
+  // Init completed
+  ctx->completed = 0;
+
+  // UCP non-blocking tag receive
+  // Inp - UCP worker, buffer, length, datatype, tag, tag mask, receive handler
+  status = ucp_tag_recv_nbx(*ucpRecvWorker,
+                           buffer,
+                           count,
+                           getTag(edge, sender),
+                           tagMask,
+                           &recvParam);
+
+  if (UCS_PTR_IS_ERR(status)) {
+    LOG(FATAL) << "Error in sending message via UCX";
+  }
+}
+
+/**
+ * UCX Send
+ * Modeled after the ISEND function of MPI
+ * @param [out] buffer - Pointer to the buffer to send
+ * @param [in] count - Size of the receiving data
+ * @param [in] ep - Endpoint to send the data to
+ * @param [out] ctx - Used for tracking the progress of the request
+ */
+void UCXChannel::UCX_Isend(const void *buffer,
+                           size_t count,
+                           ucp_ep_h ep,
+                           ucx::ucxContext* ctx) const {
+  // To hold the status of operations
+  ucs_status_ptr_t status;
+
+  // Send parameters (Mask, callback, context)
+  ucp_request_param_t sendParam;
+  sendParam.op_attr_mask = UCP_OP_ATTR_FIELD_CALLBACK |
+      UCP_OP_ATTR_FIELD_USER_DATA |
+      UCP_OP_ATTR_FLAG_NO_IMM_CMPL;
+  sendParam.cb.send = sendHandler;
+  sendParam.user_data = ctx;
+
+  // Init completed
+  ctx->completed = 0;
+  // UCP non-blocking tag send
+  // Inp - Endpoint, buffer, length, tag, send parameters
+  status = ucp_tag_send_nbx(ep,
+                            buffer,
+                            count,
+                            getTag(edge, rank),
+                            &sendParam);
+  // Check if there is an error in the request
+  if (UCS_PTR_IS_ERR(status)) {
+    LOG(FATAL) << "Error in sending message via UCX";
+  }
+  // Handle the situation where the ucp_tag_send_nbx function returns immediately
+  // without calling the send handler
+//  if (!UCS_PTR_IS_PTR(status) && status == nullptr) {
```

Comment:
remove this cmmented out code if not needed

Reply:
Okay